liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,125 @@
+        -- carefull about alpha equivalence idk if the effort is worth it.
+        go e1 e2 = do
+            -- Performing the unification under constructor is safe because 
+            -- C a₁ ... aₙ = C b₁ ... bₙ ⟺ ∀ n . a₁ = bₙ
+            EVar name1 : args1 <- pure $ flatten e1
+            EVar name2 : args2 <- pure $ flatten e2
+            guard $ name1 == name2
+            guard $ isCtor name1
+            guard $ length args1 == length args2
+            concat <$> zipWithM go args1 args2
+
+        isCtor  name = name `S.member` ctors
+        isLocal name = not (name `S.member` globals 
+                           || name `S.member` ctors
+                           || isPrefixOfSym anfPrefix name)
+
+flatten :: Expr -> [Expr]
+flatten (EApp f e) = flatten f ++ [e]
+flatten e = [e]
+
+groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
+groupUnifiableEqualities = mapMaybe (keep2 . snd) . M.groupList 
+    where -- We perform only 2-way unification
+          keep2 [e1, e2] = Just (e1, e2)
+          keep2 _        = Nothing
+
+getEqualities :: Expr -> [(Expr, Expr)]
+getEqualities (PAtom Eq e1 e2) = [(e1, e2)]
+getEqualities (PAnd eqs)       = concatMap getEqualities eqs
+getEqualities _                = []
+
+
+-- +-----------------------------------------------------+
+-- | AcyclicRewrites: collection of rewrites that cannot |
+-- | cause an infinite loop                              |
+-- +-----------------------------------------------------+
+-- This could be extracted as a separate module
+
+-- | A collection of rewrites that cannot cause an infinite loop
+newtype AcyclicRewrites = AR (M.HashMap Symbol Expr)
+
+-- We can think of the map a directed graph where every symbol is a vertex and
+-- there is an edge v₁ → v₂ if in the expression that v₁ is rewritten to v₂ is a
+-- free variable.  To guarantee that there arent any infinite loops in the
+-- rewrite procedure that will be executed by PLE, we need to esnure that there
+-- aren't any cyclic rewrites, this is ensured by adding the invariant to our
+-- graph of beeing a DAG.
+
+-- | Takes a list of rewrites and returns a list of rewrites and filters out any
+-- rewrites that would cause an infinite loop, the procedure is order biased as
+-- rewrites earlier in the list take precedence.
+unloop :: [(Symbol, Expr)] -> LocalRewrites
+unloop = LocalRewrites . toRewrites . foldl' doInsert empty 
+    where doInsert ar (s, e) = ar `fromMaybe` insert ar s e 
+
+-- | Get the "raw" list of rewrites
+toRewrites :: AcyclicRewrites -> M.HashMap Symbol Expr
+toRewrites (AR m) = m
+
+-- | Checks if there is a path from s1 to s2 O(Σ(e ∈ m) |exprSymbolSet e|) [O(m)
+-- where m is the number of edges]
+existsPath :: AcyclicRewrites -> Symbol -> Symbol -> Bool
+existsPath (AR m) s1' s2 = go s1'
+  where
+    -- Since m is a DAG, we can use DFS to check if there is a path from s1 to
+    -- s2 without worrying about infinite loops
```

Comment:
Does not change the result, but this might be slower than necessary since when there are multiple paths between two nodes in the DAG, `go` will be redundantly evaluated as many times on the same node.

Reply:
I dont get what you are suggesting

Reply:
Say the graph is
```
a -> b -> c -> d -> e
 `----->-´
```
So there is an edge between `a` and `c`.

`existsPath g 'f' 'a'` is going to visit node `c` twice, first through `a` and then through `b`. Which means that the subgraph `c -> d -> e` is also going to be scanned twice in search for `f`.

Reply:
Yo mean that we should do a BFS instead of a DFS?

Reply:
I guess that if you want to scan each node at most once, then a set of the already scanned vertices should be computed. The search strategy doesn't make a difference, I believe.

Reply:
Ahh now I got it. Yeah we could do that

Reply:
But i think is something that we should probably benchmark as usually those graphs are reasonably small and using an hash-set could be a slowdown